src/transformers/generation/utils.py

```diff
@@ -575,8 +575,11 @@ def _prepare_decoder_input_ids_for_generation(
         # no user input -> use decoder_start_token_id as decoder_input_ids
         if decoder_input_ids is None:
             decoder_input_ids = decoder_start_token_id
-        # exception: Donut checkpoints have task-specific decoder starts and don't expect a BOS token
-        elif self.config.model_type == "vision-encoder-decoder" and "donut" in self.name_or_path.lower():
+        # exception: Donut checkpoints have task-specific decoder starts and don't expect a BOS token. Note that the
+        # original checkpoints can't be detected through `self.__class__.__name__.lower()`, needing custom logic.
+        elif "donut" in self.__class__.__name__.lower() or (
+            self.config.model_type == "vision-encoder-decoder" and "donut" in self.config.encoder.model_type.lower()
```

Comment:
1st `or` clause: new checkpoints
2nd `or` clause: original donut checkpoints. `self.name_or_path` doesn't work for local files, `self.config.encoder.model_type` does

Reply:
😬